### PR TITLE
Introduce revng-artifact

### PR DESCRIPTION
**Pull request by drblallo**

### CODE REVIEW

File: include/revng/Pipes/PipelineManager.h [hunk 1]
```diff
@@ -126,7 +130,7 @@ class PipelineManager {
 
   /// A helper function used to produce all possible targets. It is used for
   /// debug purposes to see if any particular target crashes.
-  llvm::Error produceAllPossibleTargets();
+  llvm::Error produceAllPossibleTargets(bool ExpandTargets = true);
```
Comment (aleclearmind) [inline include/revng/Pipes/PipelineManager.h:130]: Mmmh, I'm not a fan of `bool` args, in particular if they have a single user, in particular if they have a default value, in particular if the default value is different from the one of the single user.

Do we really need this?
Can we make a method with a different, more eloquent name?

File: include/revng/Pipes/PipelineManager.h [hunk 2]
```diff
@@ -83,6 +83,9 @@ class PipelineManager {
   /// Path
   llvm::Error overrideModel(llvm::StringRef Path);
 
+  /// Entirelly replaces the model with the one provided
```
Comment (aleclearmind) [inline include/revng/Pipes/PipelineManager.h:83]: * Why do we still have model-specific methods in `PipelineManager`? I was hoping it could decay to be a super-regular global variable after the recent changes.
* The other method should be called `overrideModelFromFile`
* Please, pretty please, install a vim plugin spell checking comments and possibly also commit messages.
Pretty pretty please.

File: lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp [hunk 3]
```diff
@@ -2087,6 +2087,9 @@ void FunctionEntrypointAnalyzer::runInterproceduralAnalysis() {
 
 template<bool ShouldAnalyzeABI>
 bool EarlyFunctionAnalysis<ShouldAnalyzeABI>::runOnModule(Module &M) {
+  if (not M.getFunction("root") or M.getFunction("root")->isDeclaration())
+    return false;
```
Comment (aleclearmind) [inline lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp:2091]: * Are you sure we're not supposed to `assert` here?
* You seem to be using the `not getFunction(x) or M.getFunction(x)->isDeclaration()` quite a bit. Make an helper function in `IRHelpers.h`.

File: lib/Pipeline/Runner.cpp [hunk 4]
```diff
@@ -75,9 +77,9 @@ static void explainPipeline(const ContainerToTargetsMap &Targets,
   ExplanationLogger << Requirements.back().ToExecute->getName() << ":\n";
   prettyPrintStatus(Targets, ExplanationLogger, 2);
 
-  for (size_t I = Requirements.size(); I != 1; I--) {
-    StringRef StepName = Requirements[I - 2].ToExecute->getName();
-    const auto &TargetsNeeded = Requirements[I - 1].Objectives;
+  for (size_t I = Requirements.size(); I != 0; I--) {
+    StringRef StepName = Requirements[I - 1].ToExecute->getName();
+    const auto &TargetsNeeded = Requirements[I - 1].Input;
```
Comment (aleclearmind) [inline lib/Pipeline/Runner.cpp:82]: I see magic numbers changing here, which is usually scary.
Can you briefly summarize what's going on?

File: lib/Pipeline/Runner.cpp [hunk 5]
```diff
@@ -291,32 +293,33 @@ llvm::Expected<DiffMap> Runner::runAllAnalyses() {
...
   }
 
   ExplanationLogger << "PRODUCED \n";
   indent(ExplanationLogger, 1);
   ExplanationLogger << EndingStepName << ":\n";
-  CurrentContainer.enumerate().dump(ExplanationLogger, 2, false);
+  ToExec.back().ToExecute->containers().enumerate().dump(ExplanationLogger,
+                                                         2,
+                                                         false);
   ExplanationLogger << DoLog;
```
Comment (aleclearmind) [inline lib/Pipeline/Runner.cpp:323]: Wrap in `if (ExplanationLogger.isEnabled())`. Clearly isolating logging logic makes it easier to read the code.

File: lib/Pipes/FileContainer.cpp [hunk 6]
```diff
@@ -123,6 +130,12 @@ llvm::Error FileContainer::storeToDisk(llvm::StringRef Path) const {
     return llvm::Error::success();
   }
 
+  if (Path == "-") {
+    auto Buffer = unfailableGetFileAsStream(this->Path);
+
+    llvm::outs() << Buffer->getBuffer();
+  }
```
Comment (aleclearmind) [inline lib/Pipes/FileContainer.cpp:137]: Can we adopt `ToolOutputFile`? It handles `-` and does not leave around stuff in case of failure.
We're already using it [elsewhere](https://github.com/revng/revng/blob/develop/include/revng/Model/ToolHelpers.h#L235).

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: include/revng/Pipes/PipelineManager.h (hunk 2 above)
Comment (drblallo) [reply include/revng/Pipes/PipelineManager.h:83]: it was mostly to implement the -m option in tools, i can move it to main if you prefer

File: lib/Pipeline/Runner.cpp (hunk 4 above)
Comment (drblallo) [reply lib/Pipeline/Runner.cpp:82]: we were operating on outputs of the previous step to deduce inputs of the current one. The induction variable started from end and walked backward, so I -1 is the current guy, I-2 is the guy which outputs feed into the current guy.

This has been changed to operate in the space of inputs, rather than outputs that are transformed into inputs for the next step, so there is no need any longer to inspect the I-2 guy

File: include/revng/Pipes/PipelineManager.h (hunk 2 above)
Comment (aleclearmind) [reply include/revng/Pipes/PipelineManager.h:83]: Yes, makes sense.

File: lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp (hunk 3 above)
Comment (aleclearmind) [reply lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp:2091]: What scenario could trigger this?
If one asks for the binary at the end of step that includes `collect-cfg` and therefore we run the step but `cloneFilter`ing only for things we're interested in, which means, only the binary and therefore we have no `root`?

Comment (drblallo) [reply lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp:2091]: yes

File: lib/Pipes/FileContainer.cpp (hunk 6 above)
Comment (drblallo) [reply lib/Pipes/FileContainer.cpp:137]: i could but to do so it would require to load into memory the file in the general case

```
  auto Error = errorCodeToError(llvm::sys::fs::copy_file(this->Path, Path));
  auto Perm = cantFail(errorOrToExpected(fs::getPermissions(this->Path)));
  fs::setPermissions(Path, Perm);
```

Comment (aleclearmind) [reply lib/Pipes/FileContainer.cpp:137]: How does that work with `-`?

File: include/revng/Pipes/PipelineManager.h (hunk 2 above)
Comment (drblallo) [reply include/revng/Pipes/PipelineManager.h:83]: toh multiple tools need to use it to implement -m so that is a better place to keep it.

File: lib/Pipes/FileContainer.cpp (hunk 6 above)
Comment (drblallo) [reply lib/Pipes/FileContainer.cpp:137]: i load it and dump it to stdout

Comment (aleclearmind) [reply lib/Pipes/FileContainer.cpp:137]: I mean, `copy_file`, if an argument is `-` it will fail/create a file named `-`.

Comment (drblallo) [reply lib/Pipes/FileContainer.cpp:137]: yeah, that is why we have a if before everything else that check if the output path is -. In that case it executes the code in the top snippet

File: include/revng/Pipes/PipelineManager.h (hunk 1 above)
Comment (drblallo) [reply include/revng/Pipes/PipelineManager.h:130]: i can do that, but this is only used to implement --produce-all and --produce-all-single, and would introduce a copy of that function just to have a argument less. 

i would prefer less code duplication and keep it this way this time around

Comment (aleclearmind) [reply include/revng/Pipes/PipelineManager.h:130]: You can factor the implementation through an `...Impl` method.
I'd suggest to have `--produce-all-individually` and `produceAllPossibleTargetsIndividually`.
The fact that `ExpandTargets` means to produce everything individually is not very intuitive and the fact that is defaulted to `true` is quite strange too.

What's more informative for the reader?

```c++
produceAllPossibleTargets();
produceAllPossibleTargetsIndividually();
```

or

```c++
produceAllPossibleTargets(false);
produceAllPossibleTargets();
```

File: lib/Pipes/Pipes.cpp [hunk 7]
```diff
@@ -118,15 +119,55 @@ class DetectABIAnalysis {
...
+  };
+};
+
+class DetectABIAnalysis {
+private:
+  LLVMAnalysisImplementation<CollectFunctionsFromCalleesWrapperPass,
+                             EarlyFunctionAnalysis::EarlyFunctionAnalysis<true>,
+                             CollectFunctionsFromUnusedAddressesWrapperPass,
+                             EarlyFunctionAnalysis::EarlyFunctionAnalysis<true>>
+    Implementation;
```
Comment (aleclearmind) [inline lib/Pipes/Pipes.cpp:153]: `:O`

File: include/revng/Pipes/PipelineManager.h (hunk 2 above)
Comment (aleclearmind) [reply include/revng/Pipes/PipelineManager.h:83]: So you think there's a demand for an handy wrapper, gotcha.
Still he's a bit of special kid. Can we make this method non `extern` and make it a wrapper around an `overrideGlobal` method? Something like:
```c++
llvm::Error overrideModel(TupleTree<model::Binary> Binary) {
  return overrideGlobal<ModelGlobal>(ModelGlobalName, ...);
}
```

File: include/revng/Pipeline/LLVMContainer.h [hunk 8]
```diff
@@ -46,6 +46,9 @@
 
 namespace pipeline {
 
+void makeGlobalObjectsArray(llvm::Module &Module,
+                            llvm::StringRef GlobalArrayName);
```
Comment (aleclearmind) [inline include/revng/Pipeline/LLVMContainer.h:51]: What's this for? Doc

File: tests/pipeline/CMakeLists.txt [hunk 9]
```diff
@@ -172,7 +172,7 @@ function(add_invalidation_test TEST_NAME)
       ${CMAKE_BINARY_DIR}/libexec/revng/revng-pipeline \
       -P ${OPT_PIPELINE} \
       ${OPT_TARGETS} \
-      -i begin:Strings1:${OPT_INPUT} \
+      -i ${OPT_INPUT}:begin/Strings1 \
```
Comment (aleclearmind) [inline tests/pipeline/CMakeLists.txt:176]: I think this change does not belong to this commit

File: tests/pipeline/StringContainerLibrary.cpp [hunk 10]
```diff
@@ -53,7 +53,7 @@ class StringContainer : public Container<StringContainer> {
   }
 
   bool remove(const Target &Target) {
-    if (contains(Target))
+    if (not contains(Target))
```
Comment (aleclearmind) [inline tests/pipeline/StringContainerLibrary.cpp:56]: What was going on here?
Maybe this deserves its own commit.

File: tools/pipeline/artifact/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,8 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+revng_add_executable(revng-artifact Main.cpp)
+
+target_link_libraries(revng-artifact revngPipeline revngPipes revngRecompile
+                      revngEarlyFunctionAnalysis)
```
Comment (aleclearmind) [inline tools/pipeline/artifact/CMakeLists.txt:8]: Why `revngRecompile` and `revngEarlyFunctionAnalysis`? I guess we could just have `revngPipeline` here.

File: include/revng/Pipes/PipelineManager.h [hunk 12]
```diff
@@ -127,7 +127,12 @@ class PipelineManager {
 
   /// A helper function used to produce all possible targets. It is used for
   /// debug purposes to see if any particular target crashes.
-  llvm::Error produceAllPossibleTargets(bool ExpandTargets = true);
+  llvm::Error produceAllPossibleTargets() {
+    return produceAllPossibleTargets(false);
+  }
+  llvm::Error produceAllPossibleSingleTargets() {
+    return produceAllPossibleTargets(true);
+  }
```
Comment (aleclearmind) [inline include/revng/Pipes/PipelineManager.h:135]: This change doesn't seem to belong here. Can you squash it in the commit introducing `bool ExpandTargets`?

File: lib/Pipeline/Contract.cpp [hunk 13]
```diff
@@ -34,6 +34,21 @@ void Contract::deduceResults(ContainerToTargetsMap &StepStatus,
...
+  if (TargetKind) {
+    for (size_t I = TargetKind->depth(); I < In.getKind().depth(); I++) {
+      if (In.getPathComponents()[I].isSingle()) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
```
Comment (aleclearmind) [inline lib/Pipeline/Contract.cpp:50]: This is a bit obscure. Will this disappear if we get rid of `*` in targets?

If not, can you describe from an high level the problem?

File: python/revng/cli/analyze.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    def register_arguments(self, parser):
+        pass
+
+    def run(self, options: Options):
+        pipelines = collect_files(options.search_prefixes, ["share", "revng", "pipelines"], "*.yml")
+        pipelines_args = interleave(pipelines, "-P")
+        command = build_command_with_loads(
+            "revng-analyze", pipelines_args + options.remaining_args, options
+        )
+        return run(command, options)
```
Comment (aleclearmind) [inline python/revng/cli/analyze.py:22]: I guess this might be a good time to factor this class  out.

File: include/revng/Pipes/ToolCLOptions.h [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+#include <string>
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/PluginLoader.h"
+
+#include "revng/Pipes/PipelineManager.h"
+
+/*
+ * This file is distributed under the MIT License. See LICENSE.md for details.
+ */
```
Comment (aleclearmind) [inline include/revng/Pipes/ToolCLOptions.h:12]: Move right after `#pragma once`

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: tests/pipeline/StringContainerLibrary.cpp (hunk 10 above)
Comment (drblallo) [reply tests/pipeline/StringContainerLibrary.cpp:56]: that file is just a test file, it is like fixing a failing test. i can push it in a commit of it's own but there is no much point to it.

File: include/revng/Pipes/PipelineManager.h (hunk 2 above)
Comment (drblallo) [reply include/revng/Pipes/PipelineManager.h:83]: i will push it in the new CLToolOption class

File: tests/pipeline/StringContainerLibrary.cpp (hunk 10 above)
Comment (aleclearmind) [reply tests/pipeline/StringContainerLibrary.cpp:56]: Bugfixes usually deserve their own commit, unless it's a "Rework heavily this part of the code" or "Fix many bugs" commit.
But I'll leave that to you.

Comment (drblallo) [reply tests/pipeline/StringContainerLibrary.cpp:56]: it was a wrong stub implementation of a method that never got tested and now it does

File: python/revng/cli/analyze.py (hunk 14 above)
Comment (drblallo) [reply python/revng/cli/analyze.py:22]: since this affects the original pipeline tool tool i will add this change as a extra commit instead of modifying the various commits

File: include/revng/Pipeline/LLVMContainer.h (hunk 8 above)
Comment (drblallo) [reply include/revng/Pipeline/LLVMContainer.h:51]: it creates the array with a pointer to each global object so that they can't be erased by the linker.

File: lib/Pipeline/Contract.cpp (hunk 13 above)
Comment (drblallo) [reply lib/Pipeline/Contract.cpp:50]: unfortunately removing * will not help, if anything this will become even more obscure because it will have to check if all elements are present.

this is just checking that if we are going from function to root then all functions must be present.

Comment (drblallo) [reply lib/Pipeline/Contract.cpp:50]: i opted for the following comment
```
/**
 * returns false iff the output kind has less depth than the input and the input
 * is not *
 */

```

Comment (aleclearmind) [reply lib/Pipeline/Contract.cpp:50]: OK (without C-style comment `:)`).

Can we turn this into `bool Target::isCompatibleWith(Kind *Other)`? Maybe we can have a different name, but this looks like a `Targets`' method.

Comment (drblallo) [reply lib/Pipeline/Contract.cpp:50]: what do you mean c style comment? it is intended to be documentation.

it is a private method not exposed.

we can't have a compatibility thing, because it depends if it forward or backward and a couple of other things too

File: include/revng/Pipes/PipelineManager.h (hunk 12 above)
Comment (drblallo) [reply include/revng/Pipes/PipelineManager.h:135]: isolated in a comit of its own

File: lib/Pipeline/Contract.cpp (hunk 13 above)
Comment (aleclearmind) [reply lib/Pipeline/Contract.cpp:54]: Doxygen documentation in C++ uses `///`.

> we can't have a compatibility thing, because it depends if it forward or backward and a couple of other things too

Mmmh, you sure you can't make it work with `isCompatibleWith`?
`satisfiedDeeperGranularityRequest` is a very obscure name, we dropped the *granularity* term. Maybe `Contract::targetCanBePromotedTo`/`Contract::targetCanBeDemotedTo`?

Comment (drblallo) [reply lib/Pipeline/Contract.cpp:54]: targetCanBePromotedToShallowerRank?

Comment (aleclearmind) [reply lib/Pipeline/Contract.cpp:54]: `Shallower` is a big judgemental but works.

Comment (drblallo) [reply lib/Pipeline/Contract.cpp]: yeah but promote demote is unclear it it means to lower rank or higher, this is clear that means lower

File: tests/pipeline/StringContainerLibrary.cpp (hunk 10 above)
Comment (aleclearmind) [reply tests/pipeline/StringContainerLibrary.cpp:56]: Up to you, please comment whether you take action or not and mark as solved.
